File: CeliTax/SyncManager.m
//
//  SyncManager.m
//  CeliTax
//
//  Created by Leon Chen on 2015-07-23.
//  Copyright (c) 2015 CraveNSave. All rights reserved.
//

#import "SyncManager.h"
#import "SyncService.h"
#import "Utils.h"
#import "UserManager.h"

#import "CeliTax-Swift.h"

@interface SyncManager ()
{
    BOOL cancelOperations;
}

@property (nonatomic, weak) UserManager *userManager;
@property (nonatomic, weak) id <SyncService> syncService;

@property (nonatomic, strong) NSArray *filenamesToUpload;    /** Filenames that we need to upload */
@property (nonatomic, assign) NSInteger indexOfFileToUpload;

@property (nonatomic) UIBackgroundTaskIdentifier uploadImagesTask;

@property (nonatomic, strong) NSArray *filenamesToDownload;    /** Filenames that we need to download */
@property (nonatomic, strong) NSMutableArray *filenamesFailedToDownload;
@property (nonatomic, assign) NSInteger indexOfFileToDownload;
@property (nonatomic, assign) BOOL downloading;

@property (nonatomic) UIBackgroundTaskIdentifier downloadTask;

@end

@implementation SyncManager

- (instancetype) initWithSyncService: (id <SyncService>)syncService andUserManager:(UserManager *)userManager
{
    if (self = [super init])
    {
        self.syncService = syncService;
        self.userManager = userManager;
    }
    
    return self;
}

- (BOOL) needToBackUp
{
    return [self.syncService needToBackUp];
}

- (NSDate *) getLastBackUpDate
{
    return [self.syncService getLastBackUpDate];
}

- (void)checkUpdate: (NeedsUpdateBlock) needsUpdate
{
    NSString *localDataBatchID = [self.syncService getLocalDataBatchID];
    
    if (!localDataBatchID)
    {
        //no local data batch exist. Meaning the app has never been sync with server
        
        //check the server to see if the server has different data by comparing BatchID
        [self.syncService getLastestServerDataBatchID:^(NSString *batchID) {
            
            if (cancelOperations)
            {
                cancelOperations = NO;
                
                return;
            }
            
            if (!batchID)
            {
                //server has no data
                DLog(@"Server has no data to download.");
            }
            else
            {
                if (needsUpdate)
                {
                    needsUpdate();
                }
            }
            
        } failure:^(NSString *reason) {
            DLog(@"Failed to check update.");
        }];
        
    }
    //app has synced with server before
    else
    {
        //check the server to see if the server has different data by comparing BatchID
        [self.syncService getLastestServerDataBatchID:^(NSString *batchID) {
            
            if (cancelOperations)
            {
                cancelOperations = NO;
                
                return;
            }
            
            if (!batchID)
            {
                //server has no data
            }
            else
            {
                if (![localDataBatchID isEqualToString:batchID])
                {
                    if (needsUpdate)
                    {
                        needsUpdate();
                    }
                }
                
                [self downloadMissingImages];
            }
            
        } failure:^(NSString *reason) {
            
        }];
    }
}

-(void)uploadPhotos: (UploadingPhotosSuccessBlock) success
            failure: (UploadingPhotosFailureBlock) failure
{
    if (self.indexOfFileToUpload < self.filenamesToUpload.count)
    {
        NSString *filenameToUpload = (self.filenamesToUpload)[self.indexOfFileToUpload];
        
        NSData *fileData = [Utils readImageDataWithFileName: filenameToUpload forUser: self.userManager.user.userKey];
        
        if (fileData)
        {
            DLog(@"Uploading %@...", filenameToUpload);
            [self.syncService uploadFile:filenameToUpload andData:fileData success:^{
                
                DLog(@"%@ Uploaded.", filenameToUpload);
                self.indexOfFileToUpload++;
                
                [self uploadPhotos:success failure:failure];
                
            } failure:^(NSString *reason) {
                
                DLog(@"%@ failed to uploaded, stopping all uploads!", filenameToUpload);
                
                [[UIApplication sharedApplication] endBackgroundTask: self.uploadImagesTask];
                
                if (failure)
                {
                    failure(reason);
                }
                
            }];
        }
        
        //if this file doesn't exist, we have a problem with Receipt Data Integrity
        else
        {
            //Skip this file for now
            self.indexOfFileToUpload++;
            
            [self uploadPhotos:success failure:failure];
        }
    }
    else
    {
        [[UIApplication sharedApplication] endBackgroundTask: self.uploadImagesTask];
        
        if (success)
        {
            success();
        }
    }
}

- (void)startUploadingPhotos: (UploadingPhotosSuccessBlock) success
                     failure: (UploadingPhotosFailureBlock) failure
{
    //Get the list of images the server needs
    [self.syncService getFilesNeedToUpload:^(NSArray *filesnamesToUpload) {
        
        if (filesnamesToUpload.count)
        {
            //Start uploading the images one by one
            DLog(@"Need to upload:"); DLog(@"%@", filesnamesToUpload);
            
            self.filenamesToUpload = filesnamesToUpload;
            
            self.indexOfFileToUpload = 0;
            
            self.uploadImagesTask = [[UIApplication sharedApplication] beginBackgroundTaskWithExpirationHandler: nil];
            
            [self uploadPhotos:success failure:failure];
        }
        else
        {
            //Nothing to upload
            if (success)
            {
                success();
            }
        }
        
    } failure:^(NSString *reason) {
        
        if (failure)
        {
            failure(reason);
        }

    }];
}

- (void)startSync: (SyncSuccessBlock) success
          failure: (SyncFailureBlock) failure
{
    //1.Upload local data to server
    [self.syncService startSyncingUserData:^(NSDate *updateDate) {
        
        if (cancelOperations)
        {
            cancelOperations = NO;
            
            return;
        }
        
        //2. Download and merge data from server first
        [self.syncService downloadUserData:^{
            
            if (cancelOperations)
            {
                cancelOperations = NO;
                
                return;
            }
            
            //3. Delete Photos no longer attached to any receipts
            [self cleanUpReceiptImages];
            
            if (success)
            {
                success( updateDate );
            }
            
        } failure:^(NSString *reason) {
            
            if (cancelOperations)
            {
                cancelOperations = NO;
                
                return;
            }
            
            if (failure)
            {
                failure(reason);
            }
            
        }];
        
    } failure:^(NSString *reason) {
        
        if (cancelOperations)
        {
            cancelOperations = NO;
            
            return;
        }
        
        if (failure)
        {
            failure(reason);
        }
        
    }];
}

- (void)downloadAndMerge: (DownloadAndMergeDataSuccessBlock) success
                 failure: (DownloadAndMergeDataFailureBlock) failure
{
    [self.syncService downloadUserData:^{
        
        if (cancelOperations)
        {
            cancelOperations = NO;
            
            return;
        }
        
        [self downloadMissingImages];
        
        if (success)
        {
            success ();
        }
        
    } failure:^(NSString *reason) {
        
        if (cancelOperations)
        {
            cancelOperations = NO;
            
            return;
        }
        
        if (failure)
        {
            failure ( reason );
        }
        
    }];
}

-(void)downloadPhotos:(DownloadFilesSuccessBlock) success
              failure: (DownloadFileFailureBlock) failure
{
    if (cancelOperations)
    {
        self.downloading = NO;
        
        [[UIApplication sharedApplication] endBackgroundTask: self.downloadTask];
        
        cancelOperations = NO;
        
        return;
    }
    
    if (self.indexOfFileToDownload < self.filenamesToDownload.count)
    {
        NSString *filenameToDownload = (self.filenamesToDownload)[self.indexOfFileToDownload];
        
        [self.syncService downloadFile:filenameToDownload success:^{
            
            self.indexOfFileToDownload++;
            [self downloadPhotos:success failure: failure];
            
        } failure:^(NSString *reason) {
            
            [self.filenamesFailedToDownload addObject:filenameToDownload];
            
            self.indexOfFileToDownload++;
            [self downloadPhotos:success failure: failure];
            
        }];
    }
    else
    {
        self.downloading = NO;
        
        if (self.filenamesFailedToDownload.count)
        {
            if (failure)
            {
                failure (self.filenamesFailedToDownload);
            }
        }
        else
        {
            if (success)
            {
                success();
            }
        }
        
        [[UIApplication sharedApplication] endBackgroundTask: self.downloadTask];
    }
}

- (void)startDownloadPhotos:(NSArray *)filenames
                    success: (DownloadFilesSuccessBlock) success
                    failure: (DownloadFileFailureBlock) failure
{
    if (cancelOperations)
    {
        cancelOperations = NO;
        
        return;
    }
    
    if (!filenames.count)
    {
        return;
    }
    
    self.downloading = YES;
    
    self.filenamesToDownload = filenames;
    
    self.filenamesFailedToDownload = [NSMutableArray new];
    
    self.indexOfFileToDownload = 0;
    
    self.downloadTask = [[UIApplication sharedApplication] beginBackgroundTaskWithExpirationHandler: nil];
    
    [self downloadPhotos:success failure:failure];
}

- (NSArray *)getListOfFilesToDownload
{
    return [self.syncService getListOfFilesToDownload];
}

-(void) downloadMissingImages
{
    NSArray *missingImageFiles = [self getListOfFilesToDownload];
    
    if (missingImageFiles.count)
    {
        DLog(@"List of images to download: \n %@", missingImageFiles);
        [self startDownloadPhotos:missingImageFiles success:nil failure:nil];
    }
}

-(void) cleanUpReceiptImages
{
    [self.syncService cleanUpReceiptImages];
}

-(void) cancelAllOperations
{
    cancelOperations = YES;
}

@end

File: CeliTax/SyncManager.swift
//
//  SyncManager.swift
//  CeliTax
//
//  Created by Leon Chen on 2015-11-14.
//  Copyright © 2015 CraveNSave. All rights reserved.
//

import Foundation

/**
 Emcapsulates some of the more complex Syncing related functions
 */
@objc
class SyncManager : NSObject //TODO: Remove Subclass to NSObject when the entire app has been converted to Swift
{
    typealias SyncSuccessBlock = (syncDate : NSDate) -> Void
    typealias SyncFailureBlock = (reason : String) -> Void
    
    typealias UploadingPhotosSuccessBlock = () -> Void
    typealias UploadingPhotosFailureBlock = (reason : String) -> Void
    
    typealias DownloadAndMergeDataSuccessBlock = () -> Void
    typealias DownloadAndMergeDataFailureBlock = (reason : String) -> Void
    
    typealias DownloadFilesSuccessBlock = () -> Void
    typealias DownloadFileFailureBlock = (filesnamesFailedToDownload : [String]) -> Void
    
    typealias NeedsUpdateBlock = () -> Void
    typealias DoesntNeedsUpdateBlock = () -> Void
    
    private weak var userManager : UserManager!
    private weak var syncService : SyncService!
    
    private var cancelOperations : Bool = false
        
    override init()
    {
        super.init()
    }
    
    init(userManager : UserManager!, syncService : SyncService!)
    {
        self.userManager = userManager
        self.syncService = syncService
    }
    
    /*
    Insert some default categories
    */
    func insertPreloadedCategories()
    {
        self.syncService.insertPreloadedCategories();
    }
    
    /*
    Check to see if local data has a non-0 dataAction
    */
    func needToBackUp() -> Bool
    {
        return self.syncService.needToBackUp()
    }
    
    /*
    Get last Back Up Date
    */
    func getLastBackUpDate() -> NSDate?
    {
        return self.syncService.getLastBackUpDate()
    }
    
    /*
    Check if the local saved data hash ID matches the server's data's hash ID
    */
    func checkUpdate(needsUpdate : NeedsUpdateBlock?, noNeedUpdate : DoesntNeedsUpdateBlock?)
    {
        let localDataBatchID : String? = self.syncService.getLocalDataBatchID()
        
        if (localDataBatchID == nil)
        {
            //no local data batch exist. Meaning the app has never been sync with server
            
            //check the server to see if the server has different data by comparing BatchID
            self.syncService.getLastestServerDataBatchID( { (batchID) in
                
                if (self.cancelOperations)
                {
                    self.cancelOperations = false
                    
                    return
                }
                
                if (needsUpdate != nil)
                {
                    needsUpdate!()
                }
                
                }, failure: { (reason : String) in
                    dLog("Server has no data to download or failed to check update.")
                    
                    if (noNeedUpdate != nil)
                    {
                        noNeedUpdate!()
                    }
            })
        }
        else
        {
            //app has synced with server before
            //check the server to see if the server has different data by comparing BatchID
            self.syncService.getLastestServerDataBatchID( { (batchID) in
                
                if (self.cancelOperations)
                {
                    self.cancelOperations = false
                    
                    return
                }
                
                if (localDataBatchID != batchID)
                {
                    if (needsUpdate != nil)
                    {
                        needsUpdate!()
                    }
                }
                else
                {
                    if (noNeedUpdate != nil)
                    {
                        noNeedUpdate!()
                    }
                }
                
                self.downloadMissingImages()
                
                }, failure:{ (reason) in
                    //serer has no data
                    if (noNeedUpdate != nil)
                    {
                        noNeedUpdate!()
                    }
            })
        }
    }
    
    /*
    Upload any new data to server, download newest data from server, merge with local data
    */
    func startSync(success : SyncSuccessBlock?, failure : SyncFailureBlock?)
    {
        var syncDate : NSDate?
        
        dLog("Start Sync:")
        
        // Create the dispatch groups
        let serviceGroup1 : dispatch_group_t = dispatch_group_create()
        let serviceGroup2 : dispatch_group_t = dispatch_group_create()
        let serviceGroup3 : dispatch_group_t = dispatch_group_create()
    
        // Enter all groups first
        dispatch_group_enter(serviceGroup1)
        dispatch_group_enter(serviceGroup2)
        dispatch_group_enter(serviceGroup3)
        
        // 1.Upload local data to server
        
        if (self.cancelOperations)
        {
            dispatch_group_leave(serviceGroup1)
        }
        else
        {
            dLog("1.Upload local data to server")
            
            self.syncService.startSyncingUserData({ (updateDate) in
                
                dLog("Upload local data to server complete")
                
                syncDate = updateDate
                
                dispatch_group_leave(serviceGroup1)
                
            }) { (reason) in
                
                dispatch_group_leave(serviceGroup1)
                
                if (failure != nil)
                {
                    failure!(reason: reason)
                    
                    self.cancelOperations = true
                }
            }
        }
        
        // 2.Download and merge data from server
        dispatch_group_notify(serviceGroup1, dispatch_get_main_queue()) {
            
            if (self.cancelOperations)
            {
                dispatch_group_leave(serviceGroup2)
            }
            else
            {
                dLog("2.Download and merge data from server")
                
                self.syncService.downloadUserData( {
                    
                    dLog("Download and merge data from server complete")
                    
                    dispatch_group_leave(serviceGroup2)
                    
                    }, failure: { (reason) in
                        
                        dispatch_group_leave(serviceGroup2)
                        
                        if (failure != nil)
                        {
                            failure!(reason: reason)
                            
                            self.cancelOperations = true
                        }
                })
            }
            
        }
        
        // 3.Delete Photos no longer attached to any receipts
        dispatch_group_notify(serviceGroup2, dispatch_get_main_queue()) {
            
            if (self.cancelOperations)
            {
                dispatch_group_leave(serviceGroup3)
            }
            else
            {
                dLog("3.Delete Photos no longer attached to any receipts")
                
                self.cleanUpReceiptImages({ 
                    
                    dLog("Delete Photos no longer attached to any receipts complete")
                    
                    dispatch_group_leave(serviceGroup3)
                    
                })
            }
        }
        
        // Finally, trigger the success block
        dispatch_group_notify(serviceGroup3, dispatch_get_main_queue()) {
            
            if (self.cancelOperations)
            {
                self.cancelOperations = false
                
                return
            }
            
            if (success != nil)
            {
                dLog("Finally, trigger the success block")
                
                success!( syncDate: syncDate! )
            }
        }
    }
    
    /*
    Download existing data from server, merge with local data
    */
    func downloadAndMerge(success : DownloadAndMergeDataSuccessBlock?, failure : DownloadAndMergeDataFailureBlock?)
    {
        self.syncService.downloadUserData( {
            
            if (self.cancelOperations)
            {
                self.cancelOperations = false
                
                return
            }
            
            self.downloadMissingImages()
            
            if (success != nil)
            {
                success! ()
            }
            
            }, failure: { (reason) in
                
                if (self.cancelOperations)
                {
                    self.cancelOperations = false
                    
                    return
                }
                
                if (failure != nil)
                {
                    failure! ( reason: reason )
                }
                
        })
    }
    
    /*
    Secretly upload photos to server
    */
    func startUploadingPhotos(success : UploadingPhotosSuccessBlock?, failure : UploadingPhotosFailureBlock?)
    {
        // 0.Filenames that we need to upload
        var filenames : [String] = []
        
        // Create the dispatch groups
        let serviceGroup1 : dispatch_group_t = dispatch_group_create()
        
        // 1.Get the list of images the server needs
        dispatch_group_enter(serviceGroup1)
        
        self.syncService.getFilesNeedToUpload( { (filesnamesToUpload) in
            
            if (self.cancelOperations)
            {
                self.cancelOperations = false
            }
            else
            {
                if (filesnamesToUpload.count > 0)
                {
                    dLog("Need to upload:")
                    dLog(filesnamesToUpload.description)
                    
                    filenames = filesnamesToUpload
                }
            }
            
            dispatch_group_leave(serviceGroup1)
            
        }, failure: { (reason) in
            
            if (self.cancelOperations)
            {
                self.cancelOperations = false
            }
            else
            {
                failure?(reason: reason)
            }
            
            dispatch_group_leave(serviceGroup1)
        })
        
        // 2.Upload each image
        dispatch_group_notify(serviceGroup1, dispatch_get_main_queue()) {
            
            if (filenames.count == 0)
            {
                return
            }
            
            var uploadImagesTask : UIBackgroundTaskIdentifier = UIApplication.sharedApplication().beginBackgroundTaskWithExpirationHandler(nil)
            
            dLog("Upload Task started...")
            
            let uploadQueue : NSOperationQueue = NSOperationQueue()
            uploadQueue.name = "Upload queue"
            uploadQueue.maxConcurrentOperationCount = 1
            
            for fileToUpload in filenames
            {
                if (self.cancelOperations)
                {
                    self.cancelOperations = false
                    
                    UIApplication.sharedApplication().endBackgroundTask(uploadImagesTask)
                    uploadImagesTask = UIBackgroundTaskInvalid
                    
                    dLog("Upload Task complete.")
                    
                    break
                }
                
                guard let fileData : NSData = Utils.readImageDataWithFileName(fileToUpload, userKey: self.userManager.user!.userKey) else {
                    continue
                }
                
                let uploadOperation : ImageUploaderOperation = ImageUploaderOperation(syncService: self.syncService, filenameToUpload: fileToUpload, fileData: fileData)
                
                if fileToUpload == filenames.last
                {
                    uploadOperation.completionBlock = {
                        UIApplication.sharedApplication().endBackgroundTask(uploadImagesTask)
                        uploadImagesTask = UIBackgroundTaskInvalid
                        
                        dLog("Upload Task complete.")
                    }
                }
                
                uploadQueue.addOperation(uploadOperation)
            }
        }
    }
    
    /*
    Try to download the files in filenames from the server to local image storage
    */
    func startDownloadPhotos(filenames : [String], success : DownloadFilesSuccessBlock?, failure : DownloadFileFailureBlock?)
    {
        if (self.cancelOperations)
        {
            self.cancelOperations = false
            
            return
        }
        
        if (filenames.count == 0)
        {
            return
        }
        
        // Filenames that we need to download
        var downloadTask : UIBackgroundTaskIdentifier = UIApplication.sharedApplication().beginBackgroundTaskWithExpirationHandler(nil)
        
        dLog("Download Task started...")
        
        let downloadQueue : NSOperationQueue = NSOperationQueue()
        downloadQueue.name = "Download queue"
        downloadQueue.maxConcurrentOperationCount = 1
        
        for fileToDownload in filenames
        {
            if (self.cancelOperations)
            {
                self.cancelOperations = false
                
                UIApplication.sharedApplication().endBackgroundTask(downloadTask)
                downloadTask = UIBackgroundTaskInvalid
                
                dLog("Download Task complete.")
                
                break
            }
            
            let downloadOperation : ImageDownloaderOperation = ImageDownloaderOperation(syncService: self.syncService, filenameToDownload: fileToDownload)
            
            if fileToDownload == filenames.last
            {
                downloadOperation.completionBlock = {
                    UIApplication.sharedApplication().endBackgroundTask(downloadTask)
                    downloadTask = UIBackgroundTaskInvalid
                    
                    dLog("Download Task complete.")
                }
            }
            
            downloadQueue.addOperation(downloadOperation)
        }
       
    }
    
    private func getListOfFilesToDownload() -> [String]
    {
        return self.syncService.getListOfFilesToDownload()
    }
    
    /*
    Get the list of receipt images that need to be downloaded from server
    and start downloading them
    */
    func downloadMissingImages()
    {
        let missingImageFiles : [String] = self.getListOfFilesToDownload()
        
        if (missingImageFiles.count > 0)
        {
            dLog( String.init(format: "List of images to download: \n %@", missingImageFiles) )
            self.startDownloadPhotos(missingImageFiles, success:nil, failure:nil)
        }
    }
    
    /*
    Find any Photo files that are not in an exsting Receipt's filenames and delete these files
    */
    func cleanUpReceiptImages(completion : SyncService.CleanReceiptsCompletionBlock?)
    {
        self.syncService.cleanUpReceiptImages(completion)
    }
    
    func cancelAllOperations()
    {
        cancelOperations = true
    }
    
    func resetCancelOperation()
    {
        cancelOperations = false
    }
}